hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java

```diff
@@ -154,6 +154,33 @@ public void init(ClientConfig config, ClientContext clientContext) {
         }
 
         readProxyDescriptors();
+
+        this.addDistributedObjectListener(new CleanupDistObjListener(this));
+    }
+
+    /**
+     * This listener is to destroy local distributed objects when they are destroyed in the cluster.
+     * This helps to avoid memory leak. Note that member-side code uses a similar event based mechanism to
+     * create/destroy local proxies when a destroy/create occurs in the cluster. We don't create proxies in
+     * this listener because the client does not need the proxies it does not use.
+     */
+    public static class CleanupDistObjListener implements DistributedObjectListener {
+
+        private final ProxyManager proxyManager;
+
+        public CleanupDistObjListener(ProxyManager proxyManager) {
+            this.proxyManager = proxyManager;
+        }
+
+        @Override
+        public void distributedObjectCreated(DistributedObjectEvent event) {
+            // no op
+        }
+
+        @Override
+        public void distributedObjectDestroyed(DistributedObjectEvent event) {
+            proxyManager.destroyProxyLocally(event.getServiceName(), (String) event.getObjectName());
```

Comment:
Question: Let's assume I am using map "myMap" with near cache. It is destroyed at the background from local proxies and I still have the reference to the map and I make a `map.get` call. Do I get an exception in this case saying that the object was destroyed? Or it silently recreates the distributed object in the cluster, and my reference works even though there is no local proxy anymore? Is there a test for this?

Reply:
Not sure if there is a test, but if there is it did not fail. I will write try to write one.

Reply:
I know it is not an ideal test code but I wrote this test in src/test/java/com/hazelcast/map/impl/nearcache/NearCacheTest.java:


```java
class ServerThread extends Thread {
        public ServerThread() {

        }

        @Override
        public void run() {
            String mapName = "test";

            // 1. create cluster
            Config config = getConfig();
            HazelcastInstance server = Hazelcast.newHazelcastInstance(config);
            IMap<String, Integer> map = server.getMap(mapName);
            map.put("key", 1);

            assertClusterSizeEventually(1, server);

            int counter = 0;
            while (counter < 50) {
                counter++;
                if(counter == 10) {
                    map.destroy();
                }
                System.out.println(server.getDistributedObjects());
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    @Test
    public void destroying_near_cache_and_getting_it_recreates_it() {
        new ServerThread().start();
        String mapName = "test";

        // 2. add client with Near Cache
        NearCacheConfig nearCacheConfig = newNearCacheConfig();
        nearCacheConfig.setName(mapName);

        ClientConfig nearCachedConfig = new ClientConfig();
        nearCachedConfig.addNearCacheConfig(nearCacheConfig);

        HazelcastInstance client = HazelcastClient.newHazelcastClient(nearCachedConfig);
        IMap<String, Integer> map = client.getMap(mapName);

        int counter = 0;
        while (counter < 50) {
            counter++;
            map = client.getMap("test");
            System.out.println(map.get("key"));
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

    }
``` 

Since it is not following test code conventions I won't commit it. I could not write what I want with test factories etc. So I used code like this.

You can try running this code. I debugged and see that:

1. In master, client does not create a new remote map and no errors is thrown. It uses its own map in local proxies. However, there are no maps in remote, so get() returns null.
2. In my branch, client creates a new remote map and no errors is thrown. Then it is registered to local proxies.